lib/ui/recommendation.js

```diff
@@ -187,5 +197,125 @@ Recommendation.prototype = {
     this.win.setTimeout(() => {
       this.events.publish('recommendation-navigate');
     });
+  },
+  onData: function(data) {
+    this.recData = data;
+
+    // We might have both pieces of data by now; try to check for dupes.
+    this.duplicateCheck();
+  },
+  // TODO: better fn name?
+  _pollController: function() {
+    // Don't kick off more than one setTimeout loop at a time.
+    if (this.isPollingController) {
+      return;
+    }
+    this.isPollingController = true;
+
+    // Each time the urlbar changes, use a tight setTimeout loop to poll the
+    // C++ controller for results.
+    this.ctrl = Components.classes["@mozilla.org/autocomplete/controller;1"].
+                getService(Components.interfaces.nsIAutoCompleteController);
+
+    this._controllerCheck();
+  },
+  _controllerCheck: function() {
+    // searchStatus 4 means search succeeded and is done
+    if (this.ctrl.searchStatus == 4) {
```

Comment:
probably better to use the [`Ci.nsIAutoCompleteController` constants](https://dxr.mozilla.org/mozilla-central/source/toolkit/components/autocomplete/nsIAutoCompleteController.idl#15-18) than random-looking integers